src/pybind/mgr/cli_api/module.py

```diff
@@ -1,92 +1,121 @@
-        self._number_of_total_calls = number_of_total_calls
-        self._threads = []
-        self._jobs: Queue = Queue()
-        self._time = 0.0
-        self._self_time = []
-        self._lock = threading.Lock()
-
-    def start(self, func):
-        if(self._number_of_total_calls and self._number_of_parallel_calls):
-            for thread_id in range(self._number_of_parallel_calls):
-                new_thread = threading.Thread(target=ThreadedBenchmarkRunner.timer,
-                                              args=(self, self._jobs, func,))
-                self._threads.append(new_thread)
-            for job_id in range(self._number_of_total_calls):
-                self._jobs.put(job_id)
-            for thread in self._threads:
-                thread.start()
+        return {
+            'prefix': f'mgr cli {self.func.__name__}',
+            'perm': API.perm.get(self.func)
+        }
+
+
+class MgrAPIReflector(type):
+    '''
+    Metaclass to register COMMANDS and Command Handlers via CLICommand
+    decorator
+    '''
+
+    def __new__(cls, name, bases, dct):
+        klass = super().__new__(cls, name, bases, dct)
+        cls.threaded_benchmark_runner = None
+        for base in bases:
+            for name, func in inspect.getmembers(base, cls.is_public):
+                # However not necessary (CLICommand uses a registry)
+                # save functions to klass._cli_{n}() methods. This
+                # can help on unit testing
+                wrapper = cls.func_wrapper(func)
+                if API.cliapi.get(func):
+                    command = CLICommand(**CephCommander(func).to_ceph_signature())(
+                        wrapper)
+                    setattr(
+                        klass,
+                        f'_cli_{name}',
+                        command)
+        return klass
+
+    @staticmethod
+    def is_public(func: Callable) -> bool:
+        return (
+            inspect.isfunction(func)
+            and not func.__name__.startswith('_')
+            and not API.hook.get(func)
+            and not API.internal.get(func)
+        )
+
+    @staticmethod
+    def func_wrapper(func):
+        @functools.wraps(func)
+        def wrapper(self, *args, **kwargs):
+            return HandleCommandResult(
+                0,
+                json.dumps(func(self, *args, **kwargs), sort_keys=True, indent=4)
+            )
+
+        # functools doesn't change the signature when wrapping a function
+        # so we do it manually
+        signature = inspect.signature(func)
+        wrapper.__signature__ = signature
+        return wrapper
+
+
+class CLI(MgrModule, metaclass=MgrAPIReflector):
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    @CLICommand('mgr cli_benchmark')
+    def benchmark(self, iterations: int, threads: int, func_name: str, func_args: List[str] = None) -> HandleCommandResult:
+        if(iterations and threads):
```

Comment:
```suggestion
        if iterations and threads:
```
Redundant parentheses